### PR TITLE
*: add retry when clean lock failed

### PR DESCRIPTION
**Pull request by c4pt0r**
avoid phantom reads

### CODE REVIEW

**Comment by c4pt0r** [comment]
@siddontang @ngaut  PTAL

**Comment by ngaut** [comment]
Please add tests.

**Comment by c4pt0r** [comment]
the code seems a little messy, I'm going to refactor the existing code.

File: transaction.go [hunk 1]
```diff
@@ -363,13 +369,37 @@ func (txn *Txn) commitSecondaryAndCleanLock(lock *SecondaryLock, commitTs uint64
 	return nil
 }
 
+func (txn *Txn) cleanLockWithRetry(lock ThemisLock) error {
+	for i := 0; i < cleanLockMaxRetryCount; i++ {
+		if exists, err := txn.lockCleaner.isLockExisted(lock); err == nil && exists {
+			err := txn.tryToCleanLock(lock)
+			if err != nil && terror.ErrorEqual(err, ErrLockNotExpired) {
```
Comment (qiuyesuifeng) [inline transaction.go]: No need to check `err != nil`?
Seems have checked in line 381.

File: transaction.go [hunk 2]
```diff
@@ -363,13 +369,37 @@ func (txn *Txn) commitSecondaryAndCleanLock(lock *SecondaryLock, commitTs uint64
 	return nil
 }
 
+func (txn *Txn) cleanLockWithRetry(lock ThemisLock) error {
+	for i := 0; i < cleanLockMaxRetryCount; i++ {
+		if exists, err := txn.lockCleaner.isLockExisted(lock); err == nil && exists {
```
Comment (siddontang) [inline transaction.go]: we can check error first.
```
if exists, err :=; err != nil || !exists {
    return errors.Trace(err)
}

// do for existence
```

File: transaction.go [hunk 3]
```diff
@@ -51,8 +53,12 @@ type Txn struct {
 
 var (
 	// ErrSimulated is used when maybe rollback occurs error too.
-	ErrSimulated      = errors.New("Error: simulated error")
-	lockConfilctCount = 0
+	ErrSimulated           = errors.New("simulated error")
+	ErrLockNotExpired      = errors.New("lock not expired")
+	ErrCleanLockFailed     = errors.New("clean lock failed")
+	cleanLockMaxRetryCount = 30
```
Comment (qiuyesuifeng) [inline transaction.go]: How about `maxCleanLockRetryCount `?

File: transaction.go (hunk 2 above)
Comment (c4pt0r) [inline transaction.go]: !exists is not an error.

Comment (siddontang) [inline transaction.go]: yes, so we can combine them, if not error, errors.Trace(err) returns nil.

Comment (c4pt0r) [inline transaction.go]: got it, fixed in another PR

**Pull request closed without merging**